### DISCUSSION TITLE
Shortcut to run command in a floating pane

### DISCUSSION THREAD

**Discussion by pomdtr** [zellij-org/zellij#2518] [Q&A]
Hey, I'm the author of sunbeam, a command line launcher.

sunbeam pairs really well with the floating window feature of zellij.

<img width="1624" alt="image" src="https://github.com/zellij-org/zellij/assets/17577332/30ab0dd1-521e-4c10-8547-52f487313f63">

However, I was not able to setup a shortcut in my config to run sunbeam in a floating pane

```
keybinds {
  normal
    bind "Ctrl k" { Run "sunbeam"; }
}
```

Is there a way to replicate the `zellij run --floating --close-on-exit -- sunbeam` command in the config ?

**Comment by Lockszmith-GH** [comment]
from the [doc](https://zellij.dev/documentation/keybindings-possible-actions.html#possible-actions):
> Run - run the specified command
> **Required arguments**: The command to run, followed by optional arguments
> **Possible arguments**: `cwd` - current working directory, `direction` - the direction to open the new command pane
> ```kdl
>   // will run "tail -f /tmp/foo" in a pane opened below the focused one
>   bind "a" {
>       Run "tail" "-f" "foo" {
>           cwd "/tmp"
>           direction "Down"
>       }
>   }
>```

Have you tried:
```kdl
keybinds {
  normal
    bind "Ctrl k" { Run "zellij" "run" "--floating" "--close-on-exit" "--" "sunbeam"; }
}
```

Seems that should do the trick for you.

**Comment by angaz** [comment]
It seems there is quite a big difference between the docs and the code in this area. My current solution is quite a hack using features which are undocumented. It's pretty similar to what @Lockszmith-GH suggested. But a problem I had which I didn't like is that there is a left over pane which you have to press Ctrl+C to get rid of. So I looked into the code and found that the [`Run` action supports `close_on_exit`](https://github.com/zellij-org/zellij/blob/main/zellij-utils/src/kdl/mod.rs#L856), so I added that, and this still opens a pane which causes re-renders of my other panes, but it immediately closes, which is a lot nicer, but still not ideal.

```
bind "Ctrl k" {
  Run "zellij" "run" "--floating" "--close-on-exit" "--" "sunbeam" {
    close_on_exit true
  }
}
```

I'm not sure why there is such a feature gap between the CLI and the actions in the config.

**Reply by pomdtr** [reply]
Yeah I ended up going the same road. I'm looking at building a plugin to smooth out the integration between the two tools

**Reply by angaz** [reply]
If the `NewPane` action had the `command` argument, or the `Run` action had a `floating` argument, it would be really good and maybe fewer plugins would be needed. But yeah, seems like it's getting there. Unfortunately, I don't know enough Rust to contribute a refactor here.

**Comment by flinesse** [comment]
Hey, agreed with the above discussion regarding the feature discrepancy between keyboard-bindable and CLI commands, but I believe this achieves what you're looking for correct?

```
bind "Ctrl k" {
  Run "sunbeam" {
    close_on_exit true
  };
  TogglePaneEmbedOrFloating;
}
```

**Reply by pomdtr** [reply]
It's quite similar to the solution shared by angaz above. It works, but causes an unnecessary re-render of existing panes

**Comment by logoff** [comment]
Any news on this feature request? Sounds quite interesting, not only for sunbeam.

**Comment by gnur** [comment]
the solution seems quite simple in the end:

```
        bind "Super p" {
          Run "sunbeam" {
            close_on_exit true
            floating true
          }
        }
```

**Reply by angaz** [reply]
Yeah it seems like this works now, but the docs still don't show any of these possibilities: https://zellij.dev/documentation/keybindings-possible-actions#run

**Reply by joshgoebel** [reply]
Anyone figured out how to get size/width/height working with this?

**Reply by guttermonk** [reply]
@joshgoebel this worked for me. Thanks for posting the above solution @gnur.
```
            bind "Super l" {
                Run "lazygit" {
                    close_on_exit true
                    floating true
                    x "10%"
                    y "15%"
                    width "80%"
                    height "70%"
                }
            }
```

**Reply by lisiur** [reply]
How to hide all floating panes after exit when there are multiple floating panes.

**Reply by tummetott** [reply]
thank good i found this issue. Documentation would be awesome indeed

**Reply by khaitranhq** [reply]
you saved my life
